enhancements/network/cloud-egress-ip.md

```diff
@@ -0,0 +1,652 @@
+   `cloudprivateipconfig.cloud.openshift.io` CR with `.spec.items` filled out.
+4. Watches for the update of the CR and inspects `.status.items` to
+   retrieve the assignment status.
+5. Performs cluster level networking setup for the egress IPs.  
+6. Updates the `egressips.k8s.ovn.org` CR created by the user in 2. with its
+   `.status.items`, notifying the user of the final assignment status.
+
+###### openshift-sdn
+
+Note: openshift-sdn does not flag successful or unsuccessful assignment of
+egress IPs in any explicit way, it only logs unsuccessful assignment in its own
+logs. The user takes care of modifying the `hostsubnet` and `netnamespace` CRs,
+but if an incorrect setup is performed (for example by not aligning the egress
+IP with the nodes' subnet range), the user is never notified of this but needs
+to take an active action and look in the openshift-sdn logs.
+
+**Initialization**
+
+1. When openshift-sdn starts: it initializes the `hostsubnet` CR and fills in
+   the fields of each node's subnet information.  
+2. It will then continue watching nodes as to retrieve all nodes'
+   `cloud.openshift.io/private-ipconfig` (which will be updated by the cloud
+   component during its initialization) and finalize the initialization of its
+   required assignment data.
+
+**Egress IP assignment**
+
+1. User patches `hostsubnet.egressIP` / `hostsubnet.egressCIDRs`  corresponding
+   to "flagging" the nodes wished to be used for egress assignment
+2. User patches `netnamespace.egressIPs` on the namespace wished to be matched
+   for egress IP assignment
+3. openshift-sdn, having initialized all node level assignment data: performs
+   an assignment and issues a cloud egress IP assignment request by creating a
+   `cloudprivateipconfig.cloud.openshift.io` CR with `.spec.items` filled out.
+4. Watches for the update of the CR and inspects `.status.items` to
+   retrieve the assignment status.
+5. Performs cluster level networking setup for the egress IPs.  
+6. Triggers an `event` to the user notifying successful/unsuccessful assignment.
+
+Both network plugins will need to have their assignment procedure slightly
+altered when running in the cloud as to take into account the `capacity`
+information. This is left as an implementation detail depending on the network
+plugins' assignment algorithm.
+
+### Risks and Mitigations
+
+#### Risks
+
+Finalizing the work for 4.8 has a risk arising mainly CI/CD setup for this new
+component.
+
+Given future OpenShift features requiring cloud level modifications of the
+network fabric (once such existing future enhancement is: node port range
+modifications): the proposal here is to call the repository;
+openshift/cloud-network-config-controller. The name would explicitly indicate
+that this component performs cloud level network modifications, without digging
+itself into a "egress IP dedicated hole". New features would be added and
+supported by defining separate CRDs.
+
+##### Node CSR renewal
```

Comment:
This may need to be updated for the upcoming external cloud providers (eg https://github.com/kubernetes/cloud-provider-aws, https://github.com/kubernetes/cloud-provider-azure) which OCP will start using in the next few releases.

Reply:
Updated it with the upcoming out-of-tree cloud provider implementations. They are all equal to their current "legacy" implementation, so nothing should change for what concerns our concerns w.r.t the CSR approval - but I will explicitly mention that.